src/H5Tconv_complex.c

```diff
@@ -0,0 +1,2323 @@
+            if (!H5T_IS_ATOMIC(dst_p->shared->parent->shared))
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "invalid destination complex number datatype");
+            if (!src_p->shared->u.cplx.homogeneous || !dst_p->shared->u.cplx.homogeneous)
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL,
+                            "source and destination complex number datatypes must be homogeneous");
+            src_atomic = src_p->shared->parent->shared->u.atomic;
+            dst_atomic = dst_p->shared->parent->shared->u.atomic;
+            if (H5T_ORDER_LE != src_atomic.order && H5T_ORDER_BE != src_atomic.order &&
+                H5T_ORDER_VAX != src_atomic.order)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
+                            "unsupported byte order for source datatype");
+            if (H5T_ORDER_LE != dst_atomic.order && H5T_ORDER_BE != dst_atomic.order &&
+                H5T_ORDER_VAX != src_atomic.order)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
+                            "unsupported byte order for destination datatype");
+            if (dst_p->shared->size > 2 * TEMP_FLOAT_CONV_BUFFER_SIZE)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL, "destination datatype size is too large");
+            if (8 * sizeof(int64_t) - 1 < src_atomic.u.f.esize ||
+                8 * sizeof(int64_t) - 1 < dst_atomic.u.f.esize)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL, "exponent field is too large");
+            cdata->need_bkg = H5T_BKG_NO;
+
+            break;
+        }
+
+        case H5T_CONV_FREE:
+            break;
+
+        case H5T_CONV_CONV:
+            if (!src_p || !dst_p)
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a datatype");
+            if (NULL == conv_ctx)
+                HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "invalid datatype conversion context pointer");
+
+            if (H5T__conv_complex_loop(src_p, dst_p, conv_ctx, nelmts, buf_stride, buf) < 0)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_CANTCONVERT, FAIL, "unable to convert data values");
+
+            break;
+
+        default:
+            HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL, "unknown conversion command");
+    }
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5T__conv_complex() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5T__conv_complex_loop
+ *
+ * Purpose:     Implements the body of the conversion loop when converting
+ *              complex number values to another complex number type.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5T__conv_complex_loop(const H5T_t *src_p, const H5T_t *dst_p, const H5T_conv_ctx_t *conv_ctx, size_t nelmts,
```

Comment:
This function has fairly similar logic to `H5T__conv_f_f`, but some additional logic is needed to deal with the special rules around when a complex number type is +/- Infinity, NaN, 0, etc.